xmtp_mls/src/groups/key_package_cleaner_worker.rs

```diff
@@ -0,0 +1,159 @@
+use crate::client::ClientError;
+use crate::configuration::WORKER_RESTART_DELAY;
+use crate::Client;
+use futures::StreamExt;
+use std::time::Duration;
+use openmls_traits::storage::StorageProvider;
+use thiserror::Error;
+use tokio::sync::OnceCell;
+use tracing::instrument;
+use xmtp_db::{MlsProviderExt, StorageError, XmtpDb};
+use xmtp_proto::api_client::trait_impls::XmtpApi;
+use crate::identity::IdentityError;
+
+/// Interval at which the KeyPackagesCleanerWorker runs to delete expired messages.
+pub const INTERVAL_DURATION: Duration = Duration::from_secs(1);
+
+#[derive(Debug, Error)]
+pub enum KeyPackagesCleanerError {
+    #[error("storage error: {0}")]
+    Storage(#[from] StorageError),
+    #[error("client error: {0}")]
+    Client(#[from] ClientError),
+}
+
+impl KeyPackagesCleanerError {
+    fn db_needs_connection(&self) -> bool {
+        match self {
+            Self::Storage(s) => s.db_needs_connection(),
+            Self::Client(s) => s.db_needs_connection(),
+        }
+    }
+}
+
+impl<ApiClient, Db> Client<ApiClient, Db>
+where
+    ApiClient: XmtpApi + Send + Sync + 'static,
+    Db: xmtp_db::XmtpDb + 'static,
+{
+    #[instrument(level = "trace", skip_all)]
+    pub fn start_key_packages_cleaner_worker(&self) {
+        let client = self.clone();
+        tracing::trace!(
+            inbox_id = client.inbox_id(),
+            installation_id = hex::encode(client.installation_public_key()),
+            "starting key package cleaner worker"
+        );
+
+        let worker = KeyPackagesCleanerWorker::new(client);
+        worker.spawn_worker();
+    }
+}
+
+pub struct KeyPackagesCleanerWorker<ApiClient, Db> {
+    client: Client<ApiClient, Db>,
+    #[allow(dead_code)]
+    init: OnceCell<()>,
+}
+
+impl<ApiClient, Db> KeyPackagesCleanerWorker<ApiClient, Db>
+where
+    ApiClient: XmtpApi + Send + Sync + 'static,
+    Db: XmtpDb + Send + Sync + 'static,
+{
+    pub fn new(client: Client<ApiClient, Db>) -> Self {
+        Self {
+            client,
+            init: OnceCell::new(),
+        }
+    }
+    pub(crate) fn spawn_worker(mut self) {
+        xmtp_common::spawn(None, async move {
+            let inbox_id = self.client.inbox_id().to_string();
+            let installation_id = hex::encode(self.client.installation_public_key());
+            while let Err(err) = self.run().await {
```

Comment:
We should abstract this out to a worker struct / trait implementation now that we have the same pattern in 3 places.